src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,4787 @@
+	bool            *updated;
+};
+
+struct fd_dup2 {
+	int fd_src, fd_dest;
+	bool dest_closed;
+};
+
+/* working dir of current process */
+static char            cur_dir[DFS_MAX_PATH] = "";
+static bool            segv_handler_inited;
+/* Old segv handler */
+struct sigaction       old_segv;
+
+/* the flag to indicate whether initlization is finished or not */
+static bool            hook_enabled;
+static pthread_mutex_t lock_fd;
+static pthread_mutex_t lock_dirfd;
+static pthread_mutex_t lock_mmap;
+
+/* store ! umask to apply on mode when creating file to honor system umask */
+static mode_t          mode_not_umask;
+
+static int             fd_stdin = -1, fd_stdout = -1, fd_stderr = -1;
+
+static void
+finalize_dfs(void);
+static void
+update_cwd(void);
+
+struct dir_hdl {
+	d_list_t   entry;
+	dfs_obj_t *oh;
+	char       name[DFS_MAX_PATH];
+};
+
+static inline struct dir_hdl *
+hdl_obj(d_list_t *rlink)
+{
+	return container_of(rlink, struct dir_hdl, entry);
+}
+
+static bool
+key_cmp(struct d_hash_table *htable, d_list_t *rlink, const void *key, unsigned int ksize)
+{
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	return (strcmp(hdl->name, (const char *)key) == 0);
+}
+
+static void
+rec_free(struct d_hash_table *htable, d_list_t *rlink)
+{
+	int             rc;
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	rc = dfs_release(hdl->oh);
+	if (rc == ENOMEM)
+		rc = dfs_release(hdl->oh);
+	if (rc) {
+		printf("Error: dfs_release() failed: (%s)", strerror(rc));
+		exit(1);
+	}
+	free(hdl);
+}
+
+static bool
+rec_decref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	return true;
+}
+
+static uint32_t
+rec_hash(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	return d_hash_string_u32(hdl->name, strlen(hdl->name));
+}
+
+static d_hash_table_ops_t hdl_hash_ops = {.hop_key_cmp    = key_cmp,
+					  .hop_rec_decref = rec_decref,
+					  .hop_rec_free   = rec_free,
+					  .hop_rec_hash   = rec_hash};
+
+static dfs_obj_t *
+lookup_insert_dir(int idx_dfs, const char *name, mode_t *mode)
+{
+	struct dir_hdl *hdl;
+	dfs_obj_t      *oh;
+	d_list_t       *rlink;
+	size_t          len = strlen(name);
+	int             rc;
+
+	rlink = d_hash_rec_find(dfs_list[idx_dfs].dfs_dir_hash, name, len);
+	if (rlink != NULL) {
+		hdl = hdl_obj(rlink);
+		return hdl->oh;
+	}
+
+	rc = dfs_lookup(dfs_list[idx_dfs].dfs, name, O_RDWR, &oh, mode, NULL);
+	if (rc) {
+		errno = rc;
+		return NULL;
+	}
+
+	if (mode && !S_ISDIR(*mode))
+		return oh;
+
+	hdl = calloc(1, sizeof(struct dir_hdl));
+	if (hdl == NULL)
+		return NULL;
+	/* "len + 1" is used to avoid compiling issue on Ubuntu. */
+	strncpy(hdl->name, name, len + 1);
+	hdl->oh = oh;
+
+	rc = d_hash_rec_insert(dfs_list[idx_dfs].dfs_dir_hash, hdl->name, len, &hdl->entry, false);
+	if (rc) {
+		fprintf(stderr, "Failed to insert dir handle in hashtable\n");
+		rc = dfs_release(hdl->oh);
+		assert(rc == 0);
+		free(hdl);
+		return NULL;
+	}
```

Comment:
I think you need to pass true for the exclusive flag here.

If however I'm wrong then d_hash_rec_insert() does not return a failure if exclusive is false so as the code stands you could just use D_ASSERT()

Reply:
Thank you! "true" is set for "exclusive" now.